base/math.jl

```diff
@@ -1267,19 +1276,28 @@ end
         y = sign(y)*0x1.8p62
     end
     yint = unsafe_trunc(Int64, y) # This is actually safe since julia freezes the result
-    y == yint && return @noinline x^yint
-    2*xu==0 && return abs(y)*Inf*(!(y>0)) # if x==0
-    x<0 && throw_exp_domainerror(x) # |y| is small enough that y isn't an integer
-    !isfinite(x) && return x*(y>0 || isnan(x))           # x is inf or NaN
+    yisint = y == yint
+    if yisint
+        yint == 0 && return 1.0
+        use_power_by_squaring(yint) && return @noinline pow_body(x, yint)
+    end
+    2*xu==0 && return abs(y)*Inf*(!(y>0)) # if x === +0.0 or -0.0 (Inf * false === 0.0)
+    s = 1
+    if x < 0
+        !yisint && throw_exp_domainerror(x) # y isn't an integer
+        s = ifelse(isodd(yint), -1, 1)
+    end
+    !isfinite(x) && return copysign(x,s)*(y>0 || isnan(x))           # x is inf or NaN
+    return copysign(pow_body(abs(x), y), s)
+end
+
+@assume_effects :total @noinline function pow_body(x::Float64, y::Float64)
```

Comment:
I would very appreciate help to find out which effects are needed. Actually that was the first successful attempt, which rendered `^(Float64, Int64)` and `^(Float64,Float64)` foldable, as is verified in `test/math.jl` and maybe `test/numbers.jl`.
I will mitigate this assume_effects if possible.

Reply:
`@assume_effects :terminates_locally` was not enough.

Reply:
so the only problem with the effects is that `throw_exp_domainerror` was causing issues (I believe because of throw block de-optimization). However, the fix is very easy. If you just remove the LazyString, everything works out.

Reply:
> so the only problem with the effects is that throw_exp_domainerror was causing issues

Sorry to have to contradict. According to my observations, there is no problem with `throw_exp_domainerror`. Neither removing `LazyString` nor adding  or removing `@assume_effects` from there made any difference.
Also this function is not called from `pow_body(Float,Float)` directly or indirectly.

Reply:
Continuing my experiments with `@assume_effects for `pow_body(Float,Float)` lead to the result:
`@assume_effects :foldable` was sufficient to make `^(Float,Float)` foldable.

Reply:
hmm. I think I must have ran into an issue where Revise magically fixed the effects.